src/GraphQL.Analyzers/AllowedOnAnalyzer.cs

```diff
@@ -122,11 +118,28 @@ private void AnalyzeMemberAccessExpression(SyntaxNodeAnalysisContext context)
     private static void ReportFieldTypeDiagnostic(
         SyntaxNodeAnalysisContext context,
         MemberAccessExpressionSyntax memberAccessExpressionSyntax,
-        DiagnosticDescriptor diagnosticDescriptor)
+        DiagnosticDescriptor diagnosticDescriptor,
+        ImmutableArray<ITypeSymbol> allowedTypes)
     {
+        string typesString = allowedTypes.Length switch
+        {
+            1 => allowedTypes[0].Name,
+            2 => $"{allowedTypes[0].Name} or {allowedTypes[1].Name}",
+            _ => $"{string.Join(", ", allowedTypes.Take(allowedTypes.Length - 1).Select(t => t.Name))} or {allowedTypes[^1].Name}"
```

Comment:
We currently support up to 2 allowed types, but if more is needed, the analyzer is ready. Just define a new attribute with more generic arguments.